File: Components/WinForms/VB/MVVMMainDemo/Modules/Views/DataLayoutDetailCollectionViewModule.vb
Imports DevExpress.MVVM.Demos.ViewModels
Imports DevExpress.Utils.MVVM
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid

Namespace DevExpress.MVVM.Demos.Views
    Partial Public Class DataLayoutDetailCollectionViewModule
        Inherits TutorialControl

        Public Sub New()
            InitializeComponent()

            If Not mvvmContext.IsDesignMode Then
                InitBindings()
                'Here we will force first entity loading by its primary key
                MVVMContext.SetParameter(Me, CLng(2))
            End If
            LoadSourceCode(sourceCode, "InitBindings")
        End Sub
        Private Sub InitBindings()
            Dim fluent = mvvmContext.OfType(Of AccountViewModel)()
            ' We want all changes in editors to be synchronized with ViewModel state
            fluent.SetObjectDataSourceBinding(bindingSource, Function(x) x.Entity, Sub(x) x.Update())

            '			#Region "AccountTransactionDetails collection ViewModel bindings"
            'We want to show the Account's transactions collection in grid and react on this collection external changes (Reload, server-side Filtering)
            fluent.SetBinding(TransactionsGrid, Function(g) g.DataSource, Function(x) x.AccountTransactionDetails.Entities)

            transactionsGridView.Columns("Account").Visible = False
            'We want to synchronize the AccountTransactionDetails.SelectedEntity and the GridView.FocusedRowRandle in two-way manner
            fluent.WithEvent(Of GridView, FocusedRowObjectChangedEventArgs)(transactionsGridView, "FocusedRowObjectChanged").SetBinding(Function(x) x.AccountTransactionDetails.SelectedEntity, Function(args) TryCast(args.Row, Model.Transaction), Sub(gView, entity) gView.FocusedRowHandle = gView.FindRow(entity))

            'We want to proceed the Edit command when row double-clicked
            fluent.WithEvent(Of RowClickEventArgs)(transactionsGridView, "RowClick").EventToCommand(Sub(x) x.AccountTransactionDetails.Edit(Nothing), Function(x) x.AccountTransactionDetails.SelectedEntity, Function(args) (args.Clicks = 2) AndAlso (args.Button = System.Windows.Forms.MouseButtons.Left))
            'We want to proceed all the commands related to the Account's transactions collection 
            fluent.BindCommand(bbiTransaction_DetailsNew, Sub(x) x.AccountTransactionDetails.[New]())
            fluent.BindCommand(bbiTransaction_DetailsEdit, Sub(x) x.AccountTransactionDetails.Edit(Nothing), Function(x) x.AccountTransactionDetails.SelectedEntity)
            fluent.BindCommand(bbiTransaction_DetailsDelete, Sub(x) x.AccountTransactionDetails.Delete(Nothing), Function(x) x.AccountTransactionDetails.SelectedEntity)
            fluent.BindCommand(bbiTransaction_DetailsRefresh, Sub(x) x.AccountTransactionDetails.Refresh())
            'We want to show PopupMenu when row clicked by right button
            AddHandler transactionsGridView.RowClick, Sub(s, e)
                                                          If e.Clicks = 1 AndAlso e.Button = System.Windows.Forms.MouseButtons.Right Then
                                                              Transactions_DetailsPopUpMenu.ShowPopup(TransactionsGrid.PointToScreen(e.Location), s)
                                                          End If
                                                      End Sub
            '			#End Region
        End Sub
    End Class
End Namespace

File: Components/WinForms/CS/MVVMMainDemo/CodeExamples.Resources/DataLayoutDetailCollectionViewModule.cs
using DevExpress.MVVM.Demos.ViewModels;
using DevExpress.Utils.MVVM;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.MVVM.Demos.Views {
    public partial class DataLayoutDetailCollectionViewModule : TutorialControl {
        public DataLayoutDetailCollectionViewModule() {
            InitializeComponent();

            if(!mvvmContext.IsDesignMode) {
                InitBindings();
                //Here we will force first entity loading by its primary key
                MVVMContext.SetParameter(this, (long)2);
            }
            LoadSourceCode(sourceCode, "InitBindings");
        }
        void InitBindings() {
            var fluent = mvvmContext.OfType<AccountViewModel>();
            // We want all changes in editors to be synchronized with ViewModel state
            fluent.SetObjectDataSourceBinding(
                bindingSource, x => x.Entity, x => x.Update());
            
            #region AccountTransactionDetails collection ViewModel bindings
            //We want to show the Account's transactions collection in grid and react on this collection external changes (Reload, server-side Filtering)
            fluent.SetBinding(TransactionsGrid,
                g => g.DataSource, x => x.AccountTransactionDetails.Entities);
            
            transactionsGridView.Columns["Account"].Visible = false;
            //We want to synchronize the AccountTransactionDetails.SelectedEntity and the GridView.FocusedRowRandle in two-way manner
            fluent.WithEvent<GridView, FocusedRowObjectChangedEventArgs>(transactionsGridView, "FocusedRowObjectChanged")
                .SetBinding(x => x.AccountTransactionDetails.SelectedEntity,
                    args => args.Row as Model.Transaction,
                    (gView, entity) => gView.FocusedRowHandle = gView.FindRow(entity));
            
            //We want to proceed the Edit command when row double-clicked
            fluent.WithEvent<RowClickEventArgs>(transactionsGridView, "RowClick")
                         .EventToCommand(
                             x => x.AccountTransactionDetails.Edit(null), x => x.AccountTransactionDetails.SelectedEntity,
                             args => (args.Clicks == 2) && (args.Button == System.Windows.Forms.MouseButtons.Left));
            //We want to proceed all the commands related to the Account's transactions collection 
            fluent.BindCommand(bbiTransaction_DetailsNew, x => x.AccountTransactionDetails.New());
            fluent.BindCommand(bbiTransaction_DetailsEdit, x => x.AccountTransactionDetails.Edit(null), x => x.AccountTransactionDetails.SelectedEntity);
            fluent.BindCommand(bbiTransaction_DetailsDelete, x => x.AccountTransactionDetails.Delete(null), x => x.AccountTransactionDetails.SelectedEntity);
            fluent.BindCommand(bbiTransaction_DetailsRefresh, x => x.AccountTransactionDetails.Refresh());
            //We want to show PopupMenu when row clicked by right button
            transactionsGridView.RowClick += (s, e) =>
            {
                if(e.Clicks == 1 && e.Button == System.Windows.Forms.MouseButtons.Right)
                    Transactions_DetailsPopUpMenu.ShowPopup(TransactionsGrid.PointToScreen(e.Location), s);
            };
            #endregion
        }
    }
}
